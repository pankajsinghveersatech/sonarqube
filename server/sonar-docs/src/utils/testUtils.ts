/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
