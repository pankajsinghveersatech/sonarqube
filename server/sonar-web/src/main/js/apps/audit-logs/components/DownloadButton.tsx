/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { endOfDay, startOfDay, subDays } from 'date-fns';
import * as React from 'react';
import { translate } from '../../../helpers/l10n';
import { getBaseUrl } from '../../../helpers/system';
import '../style.css';
import { now, RangeOption } from '../utils';

export interface DownloadButtonProps {
  dateRange?: { from?: Date; to?: Date };
  downloadStarted: boolean;
  onStartDownload: () => void;
  selection: RangeOption;
}

const RANGE_OPTION_START = {
  [RangeOption.Today]: () => now(),
  [RangeOption.Week]: () => subDays(now(), 7),
  [RangeOption.Month]: () => subDays(now(), 30),
  [RangeOption.Trimester]: () => subDays(now(), 90)
};

const toISODateString = (date: Date) => date.toISOString();

function getRangeParams(selection: RangeOption, dateRange?: { from?: Date; to?: Date }) {
  if (selection === RangeOption.Custom) {
    // dateRange should be complete if 'custom' is selected
    if (!(dateRange?.to && dateRange?.from)) {
      return '';
    }

    return new URLSearchParams({
      from: toISODateString(startOfDay(dateRange.from)),
      to: toISODateString(endOfDay(dateRange.to))
    }).toString();
  }

  return new URLSearchParams({
    from: toISODateString(startOfDay(RANGE_OPTION_START[selection]())),
    to: toISODateString(now())
  }).toString();
}

export default function DownloadButton(props: DownloadButtonProps) {
  const { dateRange, downloadStarted, selection } = props;

  const downloadDisabled =
    downloadStarted ||
    (selection === RangeOption.Custom &&
      (dateRange?.from === undefined || dateRange?.to === undefined));

  const downloadUrl = downloadDisabled
    ? '#'
    : `${getBaseUrl()}/api/audit_logs/download?${getRangeParams(selection, dateRange)}`;

  return (
    <>
      <a
        className={classNames('button button-primary', { disabled: downloadDisabled })}
        download="audit_logs.json"
        onClick={downloadDisabled ? undefined : props.onStartDownload}
        href={downloadUrl}
        rel="noopener noreferrer"
        target="_blank">
        {translate('download_verb')}
      </a>

      {downloadStarted && (
        <div className="spacer-top">
          <p>{translate('audit_logs.download_start.sentence.1')}</p>
          <p>{translate('audit_logs.download_start.sentence.2')}</p>
          <br />
          <p>{translate('audit_logs.download_start.sentence.3')}</p>
        </div>
      )}
    </>
  );
}
