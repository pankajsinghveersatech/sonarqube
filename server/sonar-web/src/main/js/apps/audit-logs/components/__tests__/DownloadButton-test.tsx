/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { subDays } from 'date-fns';
import { shallow } from 'enzyme';
import * as React from 'react';
import { RangeOption } from '../../utils';
import DownloadButton, { DownloadButtonProps } from '../DownloadButton';

jest.mock('date-fns', () => {
  const { subDays } = jest.requireActual('date-fns');
  return {
    endOfDay: jest.fn().mockImplementation(d => d),
    startOfDay: jest.fn().mockImplementation(d => d),
    subDays
  };
});

jest.mock('../../utils', () => {
  const { HousekeepingPolicy, RangeOption } = jest.requireActual('../../utils');
  const now = new Date('2020-07-21T12:00:00Z');

  return {
    HousekeepingPolicy,
    now: jest.fn().mockReturnValue(now),
    RangeOption
  };
});

it.each([[RangeOption.Today], [RangeOption.Week], [RangeOption.Month], [RangeOption.Trimester]])(
  'should render correctly for %s',
  selection => {
    expect(shallowRender({ selection })).toMatchSnapshot('default');
  }
);

it('should render correctly for custom range', () => {
  const baseDate = new Date('2020-07-21T12:00:00Z');

  expect(shallowRender({ selection: RangeOption.Custom })).toMatchSnapshot('no dates');
  expect(
    shallowRender({
      dateRange: { from: subDays(baseDate, 2), to: baseDate },
      selection: RangeOption.Custom
    })
  ).toMatchSnapshot('with dates');
});

it('should handle download', () => {
  const onStartDownload = jest.fn();
  const wrapper = shallowRender({ onStartDownload });

  wrapper.find('a').simulate('click');
  wrapper.setProps({ downloadStarted: true });
  wrapper.find('a').simulate('click');

  expect(onStartDownload).toBeCalledTimes(1);
});

function shallowRender(props: Partial<DownloadButtonProps> = {}) {
  return shallow<DownloadButtonProps>(
    <DownloadButton
      downloadStarted={false}
      selection={RangeOption.Today}
      onStartDownload={jest.fn()}
      {...props}
    />
  );
}
