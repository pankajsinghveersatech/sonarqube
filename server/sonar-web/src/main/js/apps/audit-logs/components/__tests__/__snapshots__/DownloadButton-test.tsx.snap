// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly for 7days: default 1`] = `
<Fragment>
  <a
    className="button button-primary"
    download="audit_logs.json"
    href="/api/audit_logs/download?from=2020-07-14T12%3A00%3A00.000Z&to=2020-07-21T12%3A00%3A00.000Z"
    onClick={[MockFunction]}
    rel="noopener noreferrer"
    target="_blank"
  >
    download_verb
  </a>
</Fragment>
`;

exports[`should render correctly for 30days: default 1`] = `
<Fragment>
  <a
    className="button button-primary"
    download="audit_logs.json"
    href="/api/audit_logs/download?from=2020-06-21T12%3A00%3A00.000Z&to=2020-07-21T12%3A00%3A00.000Z"
    onClick={[MockFunction]}
    rel="noopener noreferrer"
    target="_blank"
  >
    download_verb
  </a>
</Fragment>
`;

exports[`should render correctly for 90days: default 1`] = `
<Fragment>
  <a
    className="button button-primary"
    download="audit_logs.json"
    href="/api/audit_logs/download?from=2020-04-22T12%3A00%3A00.000Z&to=2020-07-21T12%3A00%3A00.000Z"
    onClick={[MockFunction]}
    rel="noopener noreferrer"
    target="_blank"
  >
    download_verb
  </a>
</Fragment>
`;

exports[`should render correctly for custom range: no dates 1`] = `
<Fragment>
  <a
    className="button button-primary disabled"
    download="audit_logs.json"
    href="#"
    rel="noopener noreferrer"
    target="_blank"
  >
    download_verb
  </a>
</Fragment>
`;

exports[`should render correctly for custom range: with dates 1`] = `
<Fragment>
  <a
    className="button button-primary"
    download="audit_logs.json"
    href="/api/audit_logs/download?from=2020-07-19T12%3A00%3A00.000Z&to=2020-07-21T12%3A00%3A00.000Z"
    onClick={[MockFunction]}
    rel="noopener noreferrer"
    target="_blank"
  >
    download_verb
  </a>
</Fragment>
`;

exports[`should render correctly for today: default 1`] = `
<Fragment>
  <a
    className="button button-primary"
    download="audit_logs.json"
    href="/api/audit_logs/download?from=2020-07-21T12%3A00%3A00.000Z&to=2020-07-21T12%3A00%3A00.000Z"
    onClick={[MockFunction]}
    rel="noopener noreferrer"
    target="_blank"
  >
    download_verb
  </a>
</Fragment>
`;
